From 23d67dc11187cf656d331656234ddf667c8cbab5 Mon Sep 17 00:00:00 2001
From: Arunpandi Kannan <quic_arunpand@quicinc.com>
Date: Thu, 19 Oct 2023 14:41:09 +0530
Subject: [PATCH] wifi: mac80211:  Allow off-channel rx frames during ROC in
 STA mode

In sta mode mac80211 will not have bss configs before
connection. So DPP frames will be dropped, because
DPP happens before any connection with AP. So allow
DPP frames by checking whether device is in ROC while
it is configured in STA mode.

Signed-off-by: Arunpandi Kannan <quic_arunpand@quicinc.com>
---
 net/mac80211/rx.c | 18 ++++++++++++++++--
 1 file changed, 16 insertions(+), 2 deletions(-)

diff --git a/net/mac80211/rx.c b/net/mac80211/rx.c
index 81ab8af..6c2dd1d 100644
--- a/net/mac80211/rx.c
+++ b/net/mac80211/rx.c
@@ -5503,14 +5503,21 @@ static void __ieee80211_rx_handle_packet(struct ieee80211_hw *hw,
 							break;
 						}
 				}
+				if (!flag && sdata->vif.type == NL80211_IFTYPE_STATION &&
+				    sdata->vif.is_roc)
+					flag = true;
 			} else {
 				conf = &sdata->vif.bss_conf;
 
-				if (conf && conf->chandef.chan)
+				if (conf && conf->chandef.chan) {
 					if (cfg80211_get_hw_idx_by_freq(hw->wiphy, conf->chandef.chan->center_freq) ==
 					    cfg80211_get_hw_idx_by_freq(hw->wiphy, status->freq) ||
 					    sdata->vif.is_roc)
 						flag = true;
+				} else if (sdata->vif.type == NL80211_IFTYPE_STATION &&
+					   sdata->vif.is_roc) {
+					flag = true;
+				}
 			}
 		}
 
@@ -5563,14 +5570,21 @@ static void __ieee80211_rx_handle_packet(struct ieee80211_hw *hw,
 						}
 					}
 				}
+				if (!flag && prev->vif.type == NL80211_IFTYPE_STATION &&
+				    prev->vif.is_roc)
+					flag = true;
 			} else {
 				conf = &prev->vif.bss_conf;
 
-				if (conf && conf->chandef.chan)
+				if (conf && conf->chandef.chan) {
 					if (cfg80211_get_hw_idx_by_freq(hw->wiphy, conf->chandef.chan->center_freq) ==
 					    cfg80211_get_hw_idx_by_freq(hw->wiphy, status->freq) ||
 					    prev->vif.is_roc)
 						flag = true;
+				} else if (prev->vif.type == NL80211_IFTYPE_STATION &&
+					   prev->vif.is_roc) {
+					flag = true;
+				}
 			}
 
 			if (flag) {
-- 
2.17.1

